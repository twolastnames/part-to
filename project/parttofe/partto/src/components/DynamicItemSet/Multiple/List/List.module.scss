@import "../../../../App.scss";

.list {
  overflow-y: scroll;
}

.item {
  width: 100%;
  cursor: pointer;
  overflow-y: scroll;
  padding: var(--gap-medium) 0;
  font: var(--font-regular-large);
}

.list > div:first-of-type {
  border-top: var(--detail-color) 1px solid;
}

.row {
  display: flex;
  flex-direction: row;
  height: fit-content;
  border-bottom: var(--detail-color) 1px solid;
  margin-right: var(--gap-xlarge);
}
