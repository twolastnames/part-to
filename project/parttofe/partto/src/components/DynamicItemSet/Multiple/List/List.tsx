import React, {
  PropsWithChildren,
  useLayoutEffect,
  useRef,
  useState,
} from "react";

import { Carousel } from "../Carousel/Carousel";
import { ButtonProps } from "../../../Button/ButtonTypes";
import { MultipleProps } from "../MultipleTypes";

interface ShellProps extends PropsWithChildren {
  height: number;
  operations: Array<ButtonProps>;
}

function Shell({ children, height, operations }: ShellProps) {
  return <div style={{ height }}>{children}</div>;
}

export function List({ items }: MultipleProps) {
  const minimumHeight = 120;
  const domRef = useRef<HTMLDivElement>(null);
  const [{ height, perPage }, setCalculated] = useState<{
    height: number;
    perPage: number;
  }>({ height: 0, perPage: 0 });

  const calculateHeight = () => {
    if (!domRef.current) {
      return;
    }
    const { height: rectangleHeight } =
      domRef?.current?.getBoundingClientRect();
    const screenHeight = rectangleHeight;
    const count = Math.floor(screenHeight / minimumHeight);
    const height = Math.floor(screenHeight / count);

    setCalculated({
      height,
      perPage: count === 0 ? 1 : count,
    });
  };

  useLayoutEffect(() => {
    calculateHeight();
  }, [domRef]);

  const pages = [];
  if (height > 0) {
    for (let index = 0; index < items.length; index += perPage) {
      pages.push({
        view: (
          <div key={index}>
            {items.slice(index, index + perPage).map((item, key) => (
              <Shell
                key={`${index}-${key}`}
                height={height}
                operations={item.itemOperations}
              >
                {item.listView}
              </Shell>
            ))}
          </div>
        ),
      });
    }
  }

  return (
    <div data-testid="List" style={{ height: "100%" }} ref={domRef}>
      {items.length > 0 ? <Carousel items={items} /> : null}
    </div>
  );
}
