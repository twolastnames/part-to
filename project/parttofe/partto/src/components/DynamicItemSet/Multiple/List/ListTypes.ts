import { MultipleProps } from "../MultipleTypes";

export interface ListProps extends MultipleProps {
  onSelectionChanged: () => void;
}
