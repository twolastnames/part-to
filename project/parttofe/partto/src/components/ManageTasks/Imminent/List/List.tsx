import React from "react";

import classes from "./List.module.scss";
import { ImminentProps } from "../ImminentTypes";
import { useTaskGet } from "../../../../api/taskget";
import { Spinner } from "../../../Spinner/Spinner";
import { TimerString } from "../TimerString/TimerString";

export function List({ timestamp, till, duty }: ImminentProps) {
  const taskResponse = useTaskGet({ task: duty });
  return (
    <Spinner responses={[taskResponse]}>
      <div className={classes.list} data-testid="List">
        "Duty Start Time:"
        <TimerString started={timestamp} duration={till} />
        {`to "${taskResponse.data?.description}"`}
      </div>
    </Spinner>
  );
}
