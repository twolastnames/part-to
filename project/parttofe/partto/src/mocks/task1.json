{
    "ingredients": [],
    "tools": [],
    "duration": "300.0",
    "description": "saute diced green peppers and onion"
}
