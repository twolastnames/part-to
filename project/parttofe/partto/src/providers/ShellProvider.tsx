import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
