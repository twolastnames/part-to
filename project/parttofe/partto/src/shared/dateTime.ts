import { Duration, getDuration } from "./duration";

export enum Format {
  TIME,
}

export interface DateTime {
  sinceEpoch: () => number;
  toISOString: () => string;
  add: (arg: Duration) => DateTime;
  subtract: (arg: DateTime) => Duration;
  format: (format: Format) => string;
}

export function getDateTime(inDate?: Date): DateTime {
  const date = inDate || new Date();

  return {
    sinceEpoch: () => date.getTime(),
    toISOString: () => date.toISOString(),
    format: (format: Format) => {
      const timeWithZone = date
        .toLocaleString("en-US", { timeZoneName: "short" })
        .split(",")[1]
        .trim()
        .split(" ");
      const numbers = timeWithZone[0].split(":");
      numbers.pop();
      return [numbers.join(":"), timeWithZone[1]].join(" ");
    },
    add: (duration: Duration) =>
      getDateTime(new Date(date.getTime() + duration.toMilliseconds())),
    subtract: (subtrahend: DateTime) =>
      getDuration(date.getTime() - subtrahend.sinceEpoch()),
  };
}
