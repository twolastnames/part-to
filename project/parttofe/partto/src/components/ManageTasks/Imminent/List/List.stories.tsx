import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { List } from "./List";
import { ShellProvider } from "../../../../providers/ShellProvider";

const meta: Meta<typeof List> = {
  component: List,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof List>;

export const Simple: Story = {
  args: {},
};
