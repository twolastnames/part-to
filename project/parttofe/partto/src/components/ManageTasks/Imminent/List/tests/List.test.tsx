import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen, waitFor } from "@testing-library/react";
import { ShellProvider } from "../../../../../providers/ShellProvider";
import { List } from "../List";
import task1 from "../../../../../mocks/task1.json";
import { getDateTime } from "../../../../../shared/dateTime";
import { getDuration } from "../../../../../shared/duration";

test("snapshot", async () => {
  jest.useFakeTimers();
  jest.setSystemTime(new Date(Date.parse("04 Dec 2022 00:12:00 GMT")));
  fetchMock.mockResponse((request: Request) => {
    if (request.url.includes("/api/task/")) {
      return Promise.resolve(JSON.stringify(task1));
    }
    if (request.method === "POST") {
      return Promise.resolve(JSON.stringify({ runState: "someId" }));
    }
    return Promise.reject("");
  });
  render(
    <ShellProvider>
      <List
        timestamp={getDateTime(
          new Date(Date.parse("04 Dec 2022 00:12:00 GMT")),
        )}
        till={getDuration(3000)}
        duty="someId"
      />
    </ShellProvider>,
  );
  // eslint-disable-next-line testing-library/prefer-find-by
  const component = await waitFor(() => screen.getByTestId("List"));
  expect(component).toMatchSnapshot();
});
