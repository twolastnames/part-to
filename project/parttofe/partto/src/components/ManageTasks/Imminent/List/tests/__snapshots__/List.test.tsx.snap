// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="list"
  data-testid="List"
>
  Duty Start Time:
  <span>
    3 seconds
  </span>
   to "saute diced green peppers and onion"
</div>
`;
